### DISCUSSION TITLE
x-trap sometimes throws an uncaught error

### DISCUSSION THREAD

**Discussion by ajnsn** [alpinejs/alpine#2234] [7. Bugs (archived)]
When showing a dialogue together with x-trap, at least sometimes an uncaught error will be logged to the Console:

```
Uncaught Error: Your focus-trap must have at least one container with at least one tabbable node in it at all times
```

The focus is not set on the first input element and the focus trap does not work.
I am expierencing this on the latest Chrome, Firefox and Safari on Mac.

To reproduce, go to https://alpinejs.dev/plugins/trap and click the "Open dialogue" and "Close dialogue" multiple times. 

It seems that the [refactoring of x-trap to focus-trap](https://github.com/alpinejs/alpine/pull/2072) introduced the issue. Never expierenced this with the prior version.

I am not too deep into Alpine but maybe it's an issue together with x-show?
https://github.com/focus-trap/focus-trap/blob/master/index.js#L242-L250

Any clue how to prevent this?
Thank you!

**Comment by SimoTod** [comment]
it happens to me only if I use the keyboard (enter/esc/enter/esc) so I can switch quickly between them. 
Is that a real use case or just something flagged by QA? I think that a user is unlikely to click that quickly but I might be wrong.
You can still use the previous version for now if it's a problem for your users and, as you said, it's probably linked to some race conditions between x-trap and x-show.

**Comment by ajnsn** [comment]
We have a modal component which often shows the described behaviour when triggering the first open-state. The trap itself is not the problem but the then missing autofocus is an UX issue. 

I was able to migitate the issue by adding 100ms to the `setTimeout`, over here:
https://github.com/alpinejs/alpine/blob/main/packages/trap/src/index.js#L19-L21

```js
setTimeout(() => {
    trap.activate();
}, 100);
```

Works now as expected.

**Reply by SimoTod** [reply]
do you have a transition on that modal? I wonder if it happens because your modal doesn't show up immediately

**Reply by ajnsn** [reply]
Thank you for helping out here @SimoTod ! (Edit: Moving my answer to this reply)
Yes there is a transition but it's longer than 100ms, also transitioning opacity should not completely hide the element?

```html
<div
    x-show="isModalOpen"
    x-trap="isModalOpen"
    x-transition:enter-start="ease-out duration-300 opacity-0 translate-y-4 sm:translate-y-0 sm:scale-95"
    x-transition:enter-end="opacity-100 translate-y-0 sm:scale-100"
    x-transition:leave-start="ease-in duration-200 opacity-100 translate-y-0 sm:scale-100"
    x-transition:leave-end="opacity-0 translate-y-4 sm:translate-y-0 sm:scale-95"
    x-on:click.outside="isModalOpen = false"
>
```

**Reply by SimoTod** [reply]
Yeah, opacity is not related but transactions use a combination of requestAnimationFrame, microtasks, etc so the div might take a bit longer then expected to start showing up (i can't replicate easily though, is your website public?)

**Reply by ajnsn** [reply]
Thanks for taking time to replicate the issue. I tried the same with a blank CDN project and cannot replicate it as well. Maybe it's something with nesting inside a Livewire component, hm. But this issue seems to be existent, e.g. see the Apline demo. When I have time I will try to dig deeper and update this topic.

**Reply by Tim-Wils** [reply]
I can reproduce this here if I click about 10-15 times quickly  https://codepen.io/onstuimig/pen/PoKGVJN
